neutron-org/neutron-query-relayer
# feat: remove endblocker events

This PR introduces a lot of changes to the subscriber implementation.

It is related to the following PRs:

1. https://github.com/neutron-org/neutron/pull/49
2. https://github.com/neutron-org/neutron-contracts/pull/27
3. https://github.com/neutron-org/neutron-integration-tests/pull/10

**CHANGESET**

1. The subscriber is now spawned in the main() goroutine;
5. Its Subscribe() method takes a buffered `chan`, which is treated as a task queue where it writes the "tasks", a.k.a. the registered queries that the relayer needs to update;
6. NOTE: the queries task queue is buffered, which, if the buffer size is configured to a large enough value, will minimise the probability of subscriber being blocked. I decided not to use traditional queues with `sync.Cond` because it looked super ugly & made the idiomatic context-based shutdowns impossible;
7. On startup, it loads all active queries from Neutron (filtered by connection id, owner and query type);
8. The subscriber looks for Update, Remove and NewBlock events on Neutron. For Update, it retrieves the up-to-date query information from Neutron. For Remove, it removes the query from the list of active queries. For NewBlock, it checks for all events whether or not the update period is reached, and sends the query information to the tasks queue.

**HOW TO TEST THIS CODE?**

Check out the `feat/remove-endblocker-events` in neutron, contracts, relayer & integration-tests and run the `kv`, `tx` tests as usual.

> NOTE: most of the huge diff is the generated client code in `internal/subscriber/querier`.

> NOTE: this PR references a specific commit in Neutron as its dependency. It will be merged last.
